Operator IT failure

_Please use this to only for bug reports_. For questions or when you need help, you can use the [GitHub Discussions](https://github.com/kroxylicious/kroxylicious/discussions) or use the [community Slack](https://kroxylicious.slack.com) chat.

**Describe the bug**
In an IT the Deployment failed to become ready within 10 seconds
https://github.com/kroxylicious/kroxylicious/actions/runs/13753776334/job/38458043559

`ProxyReconcilerIT.testCreate` failed with:
```
2025-03-10T00:04:56.3016800Z [ERROR] io.kroxylicious.kubernetes.operator.ProxyReconcilerIT.testCreate -- Time elapsed: 22.40 s <<< ERROR!
2025-03-10T00:04:56.3017853Z org.awaitility.core.ConditionTimeoutException:
2025-03-10T00:04:56.3019390Z Condition with alias 'Deployment as expected' didn't complete within 10 seconds because assertion condition defined as a Lambda expression in io.kroxylicious.kubernetes.operator.ProxyReconcilerIT
2025-03-10T00:04:56.3020971Z [All deployment replicas should become ready] (1 failure)
2025-03-10T00:04:56.3021543Z -- failure 1 --
2025-03-10T00:04:56.3021884Z expected: null
2025-03-10T00:04:56.3023040Z  but was: 1
2025-03-10T00:04:56.3023733Z at ProxyReconcilerIT.lambda$assertDeploymentBecomesReady$2(ProxyReconcilerIT.java:143).
2025-03-10T00:04:56.3024784Z    at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
2025-03-10T00:04:56.3025787Z    at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
2025-03-10T00:04:56.3026808Z    at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
2025-03-10T00:04:56.3027776Z    at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:1160)
2025-03-10T00:04:56.3028790Z    at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:790)
2025-03-10T00:04:56.3030219Z    at io.kroxylicious.kubernetes.operator.ProxyReconcilerIT.assertDeploymentBecomesReady(ProxyReconcilerIT.java:137)
2025-03-10T00:04:56.3031775Z    at io.kroxylicious.kubernetes.operator.ProxyReconcilerIT.doCreate(ProxyReconcilerIT.java:130)
2025-03-10T00:04:56.3033279Z    at io.kroxylicious.kubernetes.operator.ProxyReconcilerIT.testCreate(ProxyReconcilerIT.java:105)
2025-03-10T00:04:56.3034368Z    at java.base/java.lang.reflect.Method.invoke(Method.java:580)
2025-03-10T00:04:56.3035127Z    at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-03-10T00:04:56.3035889Z    at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-03-10T00:04:56.3036705Z Caused by: org.assertj.core.error.AssertJMultipleFailuresError:
2025-03-10T00:04:56.3037523Z [All deployment replicas should become ready] (1 failure)
2025-03-10T00:04:56.3038063Z -- failure 1 --
2025-03-10T00:04:56.3038421Z expected: null
2025-03-10T00:04:56.3038758Z  but was: 1
2025-03-10T00:04:56.3039603Z at ProxyReconcilerIT.lambda$assertDeploymentBecomesReady$2(ProxyReconcilerIT.java:143)
2025-03-10T00:04:56.3040636Z    at io.kroxylicious.kubernetes.operator.ProxyReconcilerIT.lambda$assertDeploymentBecomesReady$3(ProxyReconcilerIT.java:142)
2025-03-10T00:04:56.3041369Z    at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
2025-03-10T00:04:56.3041946Z    at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
2025-03-10T00:04:56.3043011Z    at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
2025-03-10T00:04:56.3043675Z    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
2025-03-10T00:04:56.3044221Z    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
2025-03-10T00:04:56.3044834Z    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
2025-03-10T00:04:56.3045306Z    at java.base/java.lang.Thread.run(Thread.java:1583)

```

Looking at the logs currently emitted there's no way to determine what the state was. For testCreate it looks like:

```
2025-03-10T00:04:00.5737083Z 2025-03-10 00:04:00 <main> INFO  io.javaoperatorsdk.operator.junit.AbstractOperatorExtension:153 - Initializing integration test in namespace proxyreconcilerit-testcreate-ccb70
2025-03-10T00:04:04.5766123Z 2025-03-10 00:04:04 <main> WARN  Default ConfigurationService implementation:61 - Configuration for reconciler 'proxyreconciler' was not found. Known reconcilers: None.
2025-03-10T00:04:06.5782029Z 2025-03-10 00:04:06 <main> INFO  io.kroxylicious.kubernetes.operator.ProxyDeployment:185 - Using Kroxylicious operand image (quay.io/kroxylicious/kroxylicious:0.11.0-SNAPSHOT) from properties file /kroxylicious-image.properties on classpath
2025-03-10T00:04:06.5787328Z 2025-03-10 00:04:06 <main> INFO  io.javaoperatorsdk.operator.Operator:243 - Registered reconciler: 'proxyreconciler' for resource: 'class io.kroxylicious.kubernetes.api.v1alpha1.KafkaProxy' for namespace(s): [proxyreconcilerit-testcreate-ccb70]
2025-03-10T00:04:06.5790164Z 2025-03-10 00:04:06 <main> INFO  io.javaoperatorsdk.operator.Operator:147 - Operator SDK 4.9.7 (commit: 1f6cf2c) built on 2024-11-21T07:58:49.000+0000 starting...
2025-03-10T00:04:06.5791602Z 2025-03-10 00:04:06 <main> INFO  io.javaoperatorsdk.operator.Operator:153 - Client version: 6.13.5
2025-03-10T00:04:06.5794224Z 2025-03-10 00:04:06 <Controller Starter for: proxyreconciler> INFO  io.javaoperatorsdk.operator.processing.Controller:329 - Starting 'proxyreconciler' controller for reconciler: io.kroxylicious.kubernetes.operator.ProxyReconciler, resource: io.kroxylicious.kubernetes.api.v1alpha1.KafkaProxy
2025-03-10T00:04:06.6783401Z 2025-03-10 00:04:06 <Controller Starter for: proxyreconciler> INFO  io.javaoperatorsdk.operator.processing.Controller:341 - 'proxyreconciler' controller started
2025-03-10T00:04:06.7784552Z 2025-03-10 00:04:06 <ReconcilerExecutor-proxyreconciler-377> INFO  io.kroxylicious.kubernetes.operator.ProxyReconciler:114 - Completed reconciliation of proxyreconcilerit-testcreate-ccb70/proxy-a
2025-03-10T00:04:07.0796553Z 2025-03-10 00:04:07 <ReconcilerExecutor-proxyreconciler-382> INFO  io.kroxylicious.kubernetes.operator.ProxyReconciler:114 - Completed reconciliation of proxyreconcilerit-testcreate-ccb70/proxy-a
2025-03-10T00:04:17.0894564Z 2025-03-10 00:04:17 <main> INFO  io.javaoperatorsdk.operator.Operator:174 - Operator SDK 4.9.7 is shutting down...
```

We need to log more information on failure about the state of the resource.

--- a/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/ProxyReconcilerIT.java
+++ b/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/ProxyReconcilerIT.java
@@ -6,6 +6,7 @@
 
 package io.kroxylicious.kubernetes.operator;
 
+import java.time.Duration;
 import java.util.Base64;
 import java.util.List;
 import java.util.Map;
@@ -16,6 +17,7 @@ import java.util.stream.Collectors;
 import org.assertj.core.api.AbstractStringAssert;
 import org.assertj.core.api.Assumptions;
 import org.assertj.core.api.InstanceOfAssertFactories;
+import org.awaitility.core.ConditionFactory;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
@@ -29,6 +31,7 @@ import io.fabric8.kubernetes.api.model.Service;
 import io.fabric8.kubernetes.api.model.Volume;
 import io.fabric8.kubernetes.api.model.apps.Deployment;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.readiness.Readiness;
 import io.javaoperatorsdk.operator.junit.LocallyRunOperatorExtension;
 
 import io.kroxylicious.kubernetes.api.v1alpha1.KafkaClusterRef;
@@ -49,20 +52,21 @@ class ProxyReconcilerIT {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(ProxyReconcilerIT.class);
 
-    public static final String PROXY_A = "proxy-a";
-    public static final String PROXY_B = "proxy-b";
-    public static final String CLUSTER_FOO_REF = "fooref";
-    public static final String CLUSTER_FOO = "foo";
-    public static final String CLUSTER_FOO_BOOTSTRAP = "my-cluster-kafka-bootstrap.foo.svc.cluster.local:9092";
-    public static final String CLUSTER_BAR_REF = "barref";
-    public static final String CLUSTER_BAR = "bar";
-    public static final String CLUSTER_BAR_BOOTSTRAP = "my-cluster-kafka-bootstrap.bar.svc.cluster.local:9092";
-    public static final String NEW_BOOTSTRAP = "new-bootstrap:9092";
-    public static final String CLUSTER_BAZ = "baz";
-    public static final String CLUSTER_BAZ_REF = "bazref";
-    public static final String CLUSTER_BAZ_BOOTSTRAP = "my-cluster-kafka-bootstrap.baz.svc.cluster.local:9092";
+    private static final String PROXY_A = "proxy-a";
+    private static final String PROXY_B = "proxy-b";
+    private static final String CLUSTER_FOO_REF = "fooref";
+    private static final String CLUSTER_FOO = "foo";
+    private static final String CLUSTER_FOO_BOOTSTRAP = "my-cluster-kafka-bootstrap.foo.svc.cluster.local:9092";
+    private static final String CLUSTER_BAR_REF = "barref";
+    private static final String CLUSTER_BAR = "bar";
+    private static final String CLUSTER_BAR_BOOTSTRAP = "my-cluster-kafka-bootstrap.bar.svc.cluster.local:9092";
+    private static final String NEW_BOOTSTRAP = "new-bootstrap:9092";
+    private static final String CLUSTER_BAZ = "baz";
+    private static final String CLUSTER_BAZ_REF = "bazref";
+    private static final String CLUSTER_BAZ_BOOTSTRAP = "my-cluster-kafka-bootstrap.baz.svc.cluster.local:9092";
 
-    static KubernetesClient client;
+    private static KubernetesClient client;
+    private final ConditionFactory AWAIT = await().timeout(Duration.ofSeconds(20));
 
     @BeforeAll
     static void checkKubeAvailable() {
@@ -134,19 +138,16 @@ class ProxyReconcilerIT {
 
     private void assertDeploymentBecomesReady(KafkaProxy proxy) {
         // wait longer for initial operator image download
-        await().alias("Deployment as expected").untilAsserted(() -> {
+        AWAIT.alias("Deployment as expected").untilAsserted(() -> {
             var deployment = extension.get(Deployment.class, ProxyDeployment.deploymentName(proxy));
-            assertThat(deployment).isNotNull()
-                    .extracting(Deployment::getStatus)
+            assertThat(deployment)
                     .describedAs("All deployment replicas should become ready")
-                    .satisfies(status -> {
-                        assertThat(status.getReplicas()).isEqualTo(status.getReadyReplicas());
-                    });
+                    .returns(true, Readiness::isDeploymentReady);
         });
     }
 
     private void assertServiceTargetsProxyInstances(KafkaProxy proxy, Set<VirtualKafkaCluster> clusters) {
-        await().alias("cluster Services as expected").untilAsserted(() -> {
+        AWAIT.alias("cluster Services as expected").untilAsserted(() -> {
             for (var cluster : clusters) {
                 var service = extension.get(Service.class, ClusterService.serviceName(cluster));
                 assertThat(service).isNotNull()
@@ -159,7 +160,7 @@ class ProxyReconcilerIT {
     }
 
     private void assertDeploymentMountsConfigSecret(KafkaProxy proxy) {
-        await().alias("Deployment as expected").untilAsserted(() -> {
+        AWAIT.alias("Deployment as expected").untilAsserted(() -> {
             var deployment = extension.get(Deployment.class, ProxyDeployment.deploymentName(proxy));
             assertThat(deployment).isNotNull()
                     .extracting(dep -> dep.getSpec().getTemplate().getSpec().getVolumes(), InstanceOfAssertFactories.list(Volume.class))
@@ -170,7 +171,7 @@ class ProxyReconcilerIT {
     }
 
     private void assertProxyConfigContents(KafkaProxy cr, Set<String> contains, Set<String> notContains) {
-        await().alias("Secret as expected").untilAsserted(() -> {
+        AWAIT.alias("Secret as expected").untilAsserted(() -> {
             AbstractStringAssert<?> proxyConfig = assertThatProxyConfigFor(cr);
             if (!contains.isEmpty()) {
                 proxyConfig.contains(contains);
@@ -187,15 +188,15 @@ class ProxyReconcilerIT {
         KafkaProxy proxy = createdResources.proxy;
         extension.delete(proxy);
 
-        await().alias("Secret was deleted").untilAsserted(() -> {
+        AWAIT.alias("Secret was deleted").untilAsserted(() -> {
             var secret = extension.get(Secret.class, ProxyConfigSecret.secretName(proxy));
             assertThat(secret).isNull();
         });
-        await().alias("Deployment was deleted").untilAsserted(() -> {
+        AWAIT.alias("Deployment was deleted").untilAsserted(() -> {
             var deployment = extension.get(Deployment.class, ProxyDeployment.deploymentName(proxy));
             assertThat(deployment).isNull();
         });
-        await().alias("Services were deleted").untilAsserted(() -> {
+        AWAIT.alias("Services were deleted").untilAsserted(() -> {
             for (var cluster : createdResources.clusters) {
                 var service = extension.get(Service.class, ClusterService.serviceName(cluster));
                 assertThat(service).isNull();
@@ -212,7 +213,7 @@ class ProxyReconcilerIT {
         extension.replace(clusterRef);
 
         assertDeploymentBecomesReady(proxy);
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             var secret = extension.get(Secret.class, ProxyConfigSecret.secretName(proxy));
             assertThat(secret)
                     .isNotNull()
@@ -223,7 +224,7 @@ class ProxyReconcilerIT {
                     .contains(NEW_BOOTSTRAP);
         });
 
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             assertClusterServiceExists(proxy, CLUSTER_FOO);
             assertClusterServiceExists(proxy, CLUSTER_BAR);
         });
@@ -248,13 +249,13 @@ class ProxyReconcilerIT {
 
         // then
         assertDeploymentBecomesReady(proxy);
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             assertThatProxyConfigFor(proxy)
                     .doesNotContain(CLUSTER_FOO_BOOTSTRAP)
                     .contains(NEW_BOOTSTRAP);
         });
 
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             assertClusterServiceExists(proxy, CLUSTER_FOO);
             assertClusterServiceExists(proxy, CLUSTER_BAR);
         });
@@ -268,7 +269,7 @@ class ProxyReconcilerIT {
         KafkaProxy proxy = createdResources.proxy;
         KafkaClusterRef bazClusterRef = extension.create(clusterRef(CLUSTER_BAZ_REF, CLUSTER_BAZ_BOOTSTRAP));
         extension.create(virtualKafkaCluster(CLUSTER_BAZ, proxy, bazClusterRef));
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             var secret = extension.get(Secret.class, ProxyConfigSecret.secretName(proxy));
             assertThat(secret)
                     .isNotNull()
@@ -279,7 +280,7 @@ class ProxyReconcilerIT {
         });
         assertDeploymentBecomesReady(proxy);
 
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             assertClusterServiceExists(proxy, CLUSTER_FOO);
             assertClusterServiceExists(proxy, CLUSTER_BAR);
             assertClusterServiceExists(proxy, CLUSTER_BAZ);
@@ -293,7 +294,7 @@ class ProxyReconcilerIT {
         final var createdResources = doCreate();
         KafkaProxy proxy = createdResources.proxy;
         extension.delete(createdResources.cluster(CLUSTER_FOO));
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             var secret = extension.get(Secret.class, ProxyConfigSecret.secretName(proxy));
             assertThat(secret)
                     .isNotNull()
@@ -305,14 +306,14 @@ class ProxyReconcilerIT {
         });
         assertDeploymentBecomesReady(proxy);
 
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             var service = extension.get(Service.class, CLUSTER_FOO);
             assertThat(service)
                     .describedAs("Expect Service for cluster 'foo' to have been deleted")
                     .isNull();
         });
 
-        await().untilAsserted(() -> {
+        AWAIT.untilAsserted(() -> {
             assertClusterServiceExists(proxy, CLUSTER_BAR);
         });
         LOGGER.atInfo().log("Test finished");
@@ -398,7 +399,7 @@ class ProxyReconcilerIT {
     }
 
     private void assertClusterServiceExists(KafkaProxy proxy, String clusterName) {
-        await().alias("Service as expected").untilAsserted(() -> {
+        AWAIT.alias("Service as expected").untilAsserted(() -> {
             var service = extension.get(Service.class, clusterName);
             assertThat(service)
                     .describedAs("Expect Service for cluster '" + clusterName + "' to still exist")

